src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AbstractRazorDelegatingEndpoint.cs

```diff
@@ -0,0 +1,118 @@
+        where TDelegatedParams : class
+    {
+        private readonly DocumentContextFactory _documentContextFactory;
+        private readonly LanguageServerFeatureOptions _languageServerFeatureOptions;
+        private readonly ClientNotifierServiceBase _languageServer;
+        protected readonly ILogger Logger;
+
+        protected AbstractRazorDelegatingEndpoint(
+            DocumentContextFactory documentContextFactory,
+            LanguageServerFeatureOptions languageServerFeatureOptions,
+            ClientNotifierServiceBase languageServer,
+            ILogger logger)
+        {
+            _documentContextFactory = documentContextFactory ?? throw new ArgumentNullException(nameof(documentContextFactory));
+            _languageServerFeatureOptions = languageServerFeatureOptions ?? throw new ArgumentNullException(nameof(languageServerFeatureOptions));
+            _languageServer = languageServer ?? throw new ArgumentNullException(nameof(languageServer));
+            Logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        /// <summary>
+        /// The delegated object to send to the <see cref="CustomMessageTarget"/>
+        /// </summary>
+        protected abstract Task<TDelegatedParams> CreateDelegatedParamsAsync(TRequest request, DocumentContext documentContext, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// The name of the endpoint to delegate to, from <see cref="RazorLanguageServerCustomMessageTargets"/>. This is the
+        /// custom endpoint that is sent via <see cref="ClientNotifierServiceBase"/> which returns
+        /// a response by delegating to C#/HTML. 
+        /// </summary>
+        /// <remarks>
+        /// An example is <see cref="RazorLanguageServerCustomMessageTargets.RazorHoverEndpointName"/> 
+        /// </remarks>
+        protected abstract string CustomMessageTarget { get; }
+
+        /// <summary>
+        /// If the response needs to be handled, such as for remapping positions back, override and handle here
+        /// </summary>
+        protected virtual Task<TResponse> HandleDelegatedResponseAsync(TResponse delegatedResponse, DocumentContext documentContext, CancellationToken cancellationToken)
+            => Task.FromResult(delegatedResponse);
+
+        /// <summary>
+        /// If the request can be handled without delegation, override this to provide a response. If a null
+        /// value is returned the request will be delegated to C#/HTML servers, otherwise the response
+        /// will be used in <see cref="Handle(TRequest, CancellationToken)"/>
+        /// </summary>
+        protected virtual Task<TResponse?> TryHandleAsync(TRequest request, DocumentContext documentContext, CancellationToken cancellationToken)
+            => Task.FromResult<TResponse?>(null);
+
+        /// <summary>
+        /// Returns true if the configuration supports this operation being handled, otherwise returns false. 
```

Comment:
Not every request is gated only on single server, since that's only if it gets delegated or handled in the cshtml server. This specifically is for cases like rename where it checks if `SupportsFileManipulation` is true.

Reply:
Expanded a bit, lmk if that helps